docs: split stack (#64)
urbit/ares: docs/split-stack.md

side is higher in memory than their heap side is called an "east frame".

Computational frames are always pushed in an alternating west/east fashion (with
the first frame always being a west frame).
Thus the two stacks logically form one single computational stack.
The stack on which the current frame is located is termed the "polarity".
Thus, the current stack frame may be said to have "west" or "east" polarity.

Every stack frame undergoes two sequential phases: the execution phase, and the
cleanup phase. Ordinary operation occurs in the execution phase, while the
cleanup phase occurs if and only if the stack frame is the current frame and we
are preparing to pop it. The memory layout changes depending on which phase the
stack frame in. To keep things simple, in what follows we only describe the
execution phase, and have a separate section for the cleanup phase.

### Execution phase

Three pointers are kept in registers - the frame pointer, stack pointer, and
allocation pointer. They are arranged for west and east frames as follows:

West frame (execution phase):
|----aXXXFLLSoooAYYYs----|
East frame (execution phase):

> |----sYYYAoooSLLFXXXa----|

Reviewer: I think this diagram does more to confuse than to enlighten. My two big issues with it are:
1. It doesn't distinguish between the FP-relative slots and stack slots
2. It makes it appear as if allocation, stack, and frame pointers take up space in memory, as opposed to pointing to parts of memory

I think we'd be better served by having diagrams of the kind @drbeefsupreme made for describing nouns, and then using these notes to refer to them and explain the *why* of the design, not the what.

Author: 1. It does - that's what `X` and `L` are. But I still agree that the diagram is hard to read.
2. This is a good point

I have this in an .svg diagram already, but I'm still tweaking it since I want to use the same one to describe an example of lightweight stack usage and I keep running into issues that make it hard to read. I'll give it another go sometime this week or early next week.

Author: Here's an svg: https://github.com/urbit/ares/pull/64/commits/bdf48555292b9a39ca6c9c9f85dd29c91fa491b9

the current allocation pointer (non-inclusive) and eastward of the previous
stack pointer (inclusive).

The following sections describe how each of these pointers are utilized.

#### Frame pointer

The frame pointer is the only pointer of the three which never moves.

Each stack frame has a number of slots (each one machine word in length) in the
memory arena. There are always at least three slots. These three slots are
reserved to store the pointers for the previous frame. Further slots may be used
to store things like the return of a function or the Nock instruction being run in
that stack frame.

These slots are sometimes called "frame pointer relative slots", or FP-relative
slots. For a west stack frame, these slots are immediately west of the frame
pointer, with the first three slots west of the frame pointer being reserved to
store the pointers of the previous frame. These slots extend to the previous
allocation pointer.

We emphasize that the frame pointer does _not_ mark a boundary for the stack
frame. For a west frame, the pointers are always ordered as previous allocation
pointer is less than the current frame pointer is less than or equal to the

> current stack pointer. For an east frame, substitute "less" with "greater". The

Reviewer: >  For a west frame, the pointers are always ordered as previous allocation
pointer is less than the current frame pointer is less than or equal to the
current stack pointer.

This is a really confusing sentence. It would probably be clearer as a bullet point list. Ex:

For a west frame:
- `prev_alloc_ptr = (frame_ptr - 3)`
- `stack_ptr >= frame_ptr`
- `*(frame_ptr - 2) = prev_alloc_ptr`
- `*(frame_ptr - 1) = prev_stack_ptr`
- `*(frame_ptr) = prev_frame_ptr`

Note that this also includes information from the "Execution Phase" section as well.

While writing this, it occurred to me: why do we need a slot for the previous allocation pointer, if the contents are always `frame_ptr +/- 2`, depending on the polarity? I recall this came up early during the work, but I don't remember what the reason was.

Author: Agreed that a list of equations is more lucid.

> While writing this, it occurred to me: why do we need a slot for the previous allocation pointer, if the contents are always frame_ptr +/- 2, depending on the polarity? I recall this came up early during the work, but I don't remember what the reason was.

That's only true if there are no additional FP-slots beyond the reserved slots. Right now, it's actually

- prev_alloc_ptr = (frame_ptr - 3 - slots)

You could do this though - you could either put the saved pointers west of `frame_pointer` and additional slots east of `frame_pointer, or put the frame pointer at the westmost word of a west frame, and FP-relative slots to the east. You'd need to save the initial value of `stack_pointer` to tell when it is empty. So you no longer need to include the previous allocation pointer as you point out, but now need to include the initial value of `stack_pointer` for that frame to tell when the lightweight stack is empty.

So you use the same amount of memory, but you end up with a different interface for the programmer. I think the way it currently works ultimately has a better interface. One reason for this is as follows.

When a frame is in "cleanup phase", you have to save the parent frame's pointers in the free space adjacent to the current frame's `alloc_pointer`. Then when you start copying from the current frame's allocations to the previous frame's, you adjust the value of  `prev_alloc_pointer` as you go. So `prev_alloc_pointer` can no longer be calculated relative to the current frame's `frame_pointer`. Thus there's logical continuity going from

```(prev_alloc_pointer, prev_frame_pointer, prev_stack_pointer)` -> `(prev_alloc_pointer, prev_frame_pointer, prev_stack_pointer)```
when you go from execution to cleanup phase, vs

```(prev_frame_pointer, prev_stack_pointer, initial_stack_pointer)` -> `(prev_alloc_pointer, prev_frame_pointer, prev_stack_pointer)```
which is what you get when you have the initial value of `stack_pointer` as a saved pointer instead of `prev_alloc_pointer`.

I thought I had a second reason but I forgot by the time I got here. Just the one reason should be enough though.

Reviewer: That makes sense - I was misunderstanding how slots were tracked. Thank you for the explanation.

Every time an allocation is made or reclaimed, the allocation
pointer moves. For a west frame, an allocation moves the allocation pointer
westward. For an east frame, an allocation moves the allocation pointer eastward.

### Pushing a stack frame

When a new stack frame is pushed, the following occurs.

First, a number of slots is specified (0 or more). These are frame-pointer
relative slots beyond the 3 necessary to store the previous frame's pointers.

The new frame pointer is then set to the previous frame's allocation pointer, offset
east (for a west frame) or west (for an east frame) by 3 plus the specified
number of slots in machine words.

The new stack pointer is set to be equal to the new frame pointer.

The new allocation pointer is set to be equal to the previous frame's stack pointer.

The relative position of the stack and allocation pointers determines whether a
stack frame is called west or east. If the stack pointer is west of the 
allocation pointer, the polarity is west. If the stack pointer is east of the
allocation pointer, the polarity is east. Thus, by the process outlined about,
we see that when the current stack frame has a west polarity, the following one

> will always have an east polarity, and vice versa.

Reviewer: Since the frame pointer never moves and the allocation pointer always moves towards it, I think it would be better to explain polarity with reference to the frame and allocation pointers instead of the stack and allocation pointers.

Reviewer 2: This is a good point. @drbeefsupreme can you make this change when you get a chance please? Since we merged the split stack I'd like to get the docs updated ASAP

Author: https://github.com/urbit/ares/pull/64/commits/0e3245bfd70f2940d33b9339408ca845cd241532

## Memory Layout

Computation proceeds in a memory arena, using two stacks growing towards each
other from opposite ends. Stack frames are "split" between the two stacks into
what one might typically call a stack and heap, with each successive frame
alternating which end of the memory it uses as a stack and which end it uses as
a heap. This looks a lot like Vere's road system, though while Vere utilizes
roads for virtualization, Ares utilizes this system for tail call optimization.
That being said, we do not presume any knowledge of Vere's road system in this
document.

Lower memory is directionally "west", and higher memory is directionally "east".
A frame whose "stack side" (to be defined shortly) is lower in memory than their
"heap side" is called a "west frame". On the other hand, a frame whose stack
side is higher in memory than their heap side is called an "east frame".

Computational frames are always pushed in an alternating west/east fashion (with
the first frame always being a west frame).
Thus the two stacks logically form one single computational stack.
The stack on which the stack side of the current frame is located is termed the "polarity".

> Thus, the current stack frame may be said to have "west" or "east" polarity.

Reviewer 2: :+1:

 - All stack frames are in the execution phase, or
 - The top stack frame is in the cleanup phase, and all others are in the
   execution phase.

The memory layout changes depending on which phase the
stack frame in. To keep things simple, in what follows we only describe the
execution phase, and have a separate section for the cleanup phase.

### Execution phase

Three pointers are kept in registers - the frame pointer, stack pointer, and
allocation pointer. I strongly recommend looking at the diagram split-stack.svg
in this folder - a simple text diagram does not convey this information well.

They are arranged for west and east frames as follows.

West frame (execution phase):
|----aXXXFLLSoooAYYYs----|
East frame (execution phase):
|----sYYYAoooSLLFXXXa----|

key:
- senior memory
A current allocation pointer
a previous allocation pointer
F current frame pointer
f previous frame pointer
S current stack pointer
s previous stack pointer
X Current memory slots
L Current lightweight stack
Y Current allocations
o free memory

#### West and East frames

For a west frame with `n` slots, the pointers are located in memory as follows:

 - `frame_ptr <= stack_ptr < alloc_ptr`

>  - `prev_alloc_ptr = frame_ptr - 3 - n`

Reviewer: Scratch that - forgot that we're talking about the pointer, not the value *at* the pointer

`prev_frame_ptr`). Another way to think about this is that the beginning of a
machine word is always its westmost bit, so there's a fundamental asymmetry in
how memory is laid out for west and east frames that manifests as the offset difference.
 
That being said, the interface is set up so the programmer does not need to
think about this. They never need to be aware of or need to check whether the current
stack frame is east or west - they will call the same functions either way. See
the [interface](#interface) section for more details.
 
The "stack side" for a west frame is the memory eastward of the previous
allocation pointer (inclusive) and west of the current stack pointer
(non-inclusive). The "heap side" for a west frame is the memory eastward of the
allocation pointer (inclusive) and westward of the previous stack pointer
(non-inclusive).

Similarly, the stack side for an east frame is the memory westward of the
previous allocation pointer (non-inclusive) and eastward of the current stack
pointer (inclusive). The heap side for an east frame is the memory westward of
the current allocation pointer (non-inclusive) and eastward of the previous

> stack pointer (inclusive).

Reviewer: If my above comment is applied, the concepts of "stack side" and "heap side" have already been explained. I also feel that this part is a little unclear, in that the slots and frame pointers are lumped in with the stack. If this section is kept at all, it might be clearer to have something like the following:
```
The stack for a west frame is the memory eastward of the frame pointer
(inclusive) and westward of the current stack pointer (exclusive). The heap 
or a west frame is the memory eastward of the current allocation pointer
(inclusive) and westward of the previous stack pointer (exclusive).

Similarly, the stack for an east frame is the memory westward of the frame
pointer (exclusive) and eastward of the current stack pointer (inclusive).
The heap for an east frame is the memory westward of the current allocation 
pointer (exclusive) and eastward of the previous stack pointer (inclusive). 

The slots and registers for a frame are always stored in the memory between
the previous allocation pointer and the frame pointer (inclusive).
```

Reviewer 2: there's no registers in the stack. There are "registers" (fields of the NockStack struct) which define the current state of the stack, but words referenced by a fixed offset from the frame pointer are called "slots"

Reviewer: Updated comment to remove calling the three permanent slots "registers"

The following sections describe how each of these pointers are utilized.

#### Frame pointer

The frame pointer is the only pointer of the three which never moves.

Each stack frame has a number of slots (each one machine word in length) in the
memory arena. There are always at least three slots. These three slots are
reserved to store the pointers for the previous frame. Further slots may be used
to store things like the return of a function or the Nock instruction being run in

> that stack frame.

Reviewer: Above, the three reserved slots are referred to as "registers"

Reviewer 2: The only thing which is correct to call a "register" in this document is the *current* stack/frame/allocation pointers.

These slots are sometimes called "frame pointer relative slots", or FP-relative
slots. For a west stack frame, these slots are immediately west of the frame
pointer, with the first three slots west of the frame pointer being reserved to
store the pointers of the previous frame. These slots extend to the previous
allocation pointer.

We emphasize that, unlike many common stack schemas, the frame pointer does
_not_ mark a boundary for the stack frame. For a west frame, the memory west of
`frame_ptr` are slots in which other pointers are stored. If the lightweight
stack is non-empty, then the memory east of `frame_ptr` contains elements of the
lightweight stack.

The ultimate reason for this is that it improves the ergonomics of the pointer
interface. A deeper explanation is given [at the end](#frame-boundary), since

> you need to understand what all the pointers are for and what the cleanup phase

Reviewer: "... you need to **first** understand ... "

"lightweight stack", which is a stack entirely contained within the current
stack frame. It is utilized for several things, including jam, cue, mug,
unifying equality, noun copying, and executing subformulas of a Nock formula in
the current stack frame. The primary reason for the existence of the lightweight
stack is to enable completely reliable tail calls. Indirect calls, which may
need arbitrarily many stack slots, utilize the lightweight stack for this
purpose.

For a west frame, pushing onto the lightweight stack moves the stack pointer
eastward, while popping moves it westward. For an east frame, pushing onto the
lightweight stack moves the stack pointer westward, while popping moves it
eastward.

#### Allocation pointer

The allocation pointer is always initialized to be equal to the previous frame's
stack pointer when a stack frame is pushed. This pointer is used to represent
the extent of the "heap" for the current stack frame. The remaining free memory
at a moment in time for the virtual machine is always precisely the region
between the current stack frame's stack pointer and allocation pointer (at least
during the execution phase). Thus the term "split stack".

> Every time an allocation is made or reclaimed, the allocation

Reviewer 2: It's because objects live in the "heap"/allocation area. The lightweight stack and stack slots only hold fixed-size references or datums. The allocation area is for arbitrary-sized nouns (and bytecode...and HAMTs...and...)

pointer moves. For a west frame, an allocation moves the allocation pointer
westward. For an east frame, an allocation moves the allocation pointer eastward.

### Pushing a stack frame

When a new stack frame is pushed, the following occurs.

First, a number of slots is specified (0 or more). These are frame-pointer
relative slots beyond the 3 necessary to store the parent frame's reserved pointers.

The new frame pointer is then set to the parent frame's allocation pointer, offset
east (for a west frame) or west (for an east frame) by 3 plus the specified
number of slots in machine words.

The new stack pointer is set to be equal to the new frame pointer.

The new allocation pointer is set to be equal to the parent frame's stack pointer.

The relative position of the frame and allocation pointers determines whether a
stack frame is called west or east. If the frame pointer is west of the 
allocation pointer, the polarity is west. If the frame pointer is east of the

> allocation pointer, the polarity is east. Thus, by the process outlined about,

Reviewer: "... the process outlined **above** ..."

The cleanup phase is the second phase a stack frame undergoes, and occurs only
when it is the current (youngest) stack frame and is preparing to be popped. In
order to sidestep the problem that copying allocations in the current frame's
allocation arena into the previous frame's will overwrite the FP-relative slots,
we copy the slots reserved for the previous frame's pointers into the free
memory adjacent to the current allocation arena. Furthermore, we also move the
lightweight stack (used to traverse nouns being copied) adjacent to that. After
the previous frame's pointers have been copied to their new location and the
stack pointer has moved, but before any copying of allocations has begun, the
memory layout looks as follows. As before, I strongly recommend looking at the
diagram split-stack.svg in this folder.

West frame (cleanup phase):
|----aXXXFoooSXXXAYYYs---|
East frame (cleanup phase):
|----sYYYAXXXSoooFXXXa---|

key:
- senior memory
A current allocation pointer
a previous allocation pointer
F current frame pointer
f previous frame pointer
S current stack pointer
s previous stack pointer
X Current memory slots
L Current lightweight stack
Y Current allocations

> o free memory

Reviewer: See above: the SVG is so much better that this ASCII diagram should be axed and the reader directed to the visual reference

Reviewer 2: Agreed. If the SVG is committed it could even be included as a markdown image reference:

```markdown
![alt text](relative_url.svg)
```

Reviewer: Or an `img` tag so that we can control the width/height of the image, plus link to the original

cleanup phase - see [interface](#interface).

#### West and East frames

For a west frame, the pointers are located in memory as follows:

 - `frame_ptr < stack_ptr < alloc_ptr`
 - `*(alloc_ptr - 1) = prev_frame_ptr`
 - `*(alloc_ptr - 2) = prev_stack_ptr`
 - `*(alloc_ptr - 3) = prev_alloc_ptr`
 - The initial position of the `stack_ptr` is `alloc_ptr - 4`.
 - `frame_ptr` and `alloc_ptr` are frozen, `stack_ptr` can move.
 - Pushing the lightweight stack moves `stack_ptr` westward.

 For an east frame, the pointers are located in memory as follows:
 - `alloc_ptr < stack_ptr > frame_ptr`
 - `*alloc_ptr = prev_frame_ptr`
 - `*(alloc_ptr + 1) = prev_stack_ptr`
 - `*(alloc_ptr + 2) = prev_alloc_ptr`
 - The initial position of the `stack_ptr` is `alloc_ptr + 3`.
 - `frame_ptr` and `alloc_ptr` are frozen, `stack_ptr` can move.
 - Pushing the lightweight stack moves `stack_ptr` eastward.

#### Copying

After the frame has moved into the cleanup phase, copying from the current
allocation arena to the previous one can begin.

When popping a frame, the result register is examined to provide a collection
root. Pointers which fall within the current frame's allocation arena are

> recursively chased and copied to the parent frame's allocation arena. This is

Reviewer: See above.
```
Pointers to result components which fall within the current frame are tracked down recursively from the root and copied to the parent frame's stack.
```

facilitated using the lightweight stack - for example, when a cell is found, its
head and tail pointers are pushed onto the lightweight stack. Then when we pop, say, the
head, we check to see if its located within the current stack frame or a senior
one. If it is the current stack frame we copy it to the previous stack frame,
otherwise it is left alone.

This ensures that data allocated in this frame which is live from the returned
result is not lost when the frame is popped.

The reason that this works is due to the following invariant: junior memory may

> only ever point to other memory within in the same stack frame, or more senior

Reviewer: ```suggestion
only ever point to other memory within the same stack frame, or more senior
```

memory. Senior memory may never point into junior memory.

> Copying a cell results in a new cell allocation to the parent frame's allocation
> arena, and pushing both constituent noun pointers onto the lightweight stack with the

Reviewer: ```suggestion
Copying a cell results in a new cell allocation to the parent frame's
stack, and pushing both constituent noun pointers onto the lightweight stack with the
```

 - `pre_copy()` - idempotent function that changes the current frame from
   execution to cleanup phase.
 - `prev_frame_pointer_pointer() -> *mut *mut u64` - pointer to where the previous frame pointer
   is saved in a frame.
 - `prev_stack_pointer_pointer() -> *mut *mut u64` - pointer to where the previous stack pointer
   is saved in a frame.
 - `prev_alloc_pointer_pointer() -> *mut *mut u64` - pointer to where the previous stack pointer
   is saved in a frame.
 - `is_west() -> bool` - `true` when the current stack frame is west.

#### Lightweight stack

We provide here a brief description of the key `NockStack` methods surrounding
the lightweight stack: 

 - `push<T>() -> *mut T` - push a `T` onto the lightweight stack.
 - `pop<T>()` - pop a `T` off the lightweight stack.
 - `top<T>() -> *mut T` - peek at a `T` on the top of the lightweight stack.
 - `stack_is_empty() -> bool` - true when lightweight stack is empty.

### Why `frame_ptr` does not mark a frame boundary {#frame-boundary}

Let's instead consider what would happen if `frame_ptr` for a west frame were
the westmost word of the frame - i.e. it marks a boundary for the stack frame,

> and in fact is then equal to `prev_alloc_ptr`. This should mean that you now only need to reserve 2 slots
> in each stack frame for the parent frame's pointers.

Reviewer: ```suggestion
and in fact is then equal to `prev_alloc_ptr`. This means that only two slots need to be reserved
in each stack frame for the parent frame's pointers.
```

> However, before we could test whether the lightweight stack was empty by
> computing `stack_ptr == frame_ptr`. Now instead, we will need to save the
> initial location of `stack_ptr` in a reserved slot. This is because it is not a
> fixed offset from `frame_ptr` since the total number of slots in a given stack frame
> can vary. So we end up with the same number of reserved slots in the execution
> phase by putting the frame pointer at the boundary. If this is where it ended, this would arguably
> have been a better choice.

Reviewer: ```suggestion
Now we can no longer test whether the lightweight stack is empty by computing `stack_ptr == frame_ptr`. Instead, we need to save the
initial location of `stack_ptr` in a reserved slot. This is because it is not a
fixed offset from `frame_ptr` since the total number of slots in a given stack frame
can vary. Therefore, we have the same number of reserved slots in the execution phase.
```

However, recall that in the cleanup phase, the reserved pointers for the parent
frame are moved from being `frame_ptr`-relative slots to `alloc_ptr`-relative
slots. So `prev_alloc_ptr` can no longer be calculated relative to
`frame_ptr`, and instead needs to be saved in an `alloc_ptr`-relative slot.

> In the cleanup phase, since the number of slots is fixed at 3 (as opposed to
> varying as in the execution phase), the emptiness of the lightweight stack can

Reviewer: ```suggestion
In the cleanup phase, since the number of slots is fixed at 3 (as opposed to
the varying number in the execution phase), the emptiness of the lightweight stack can
```

> be determined by checking if its equal to one of the reserved slots (which one
> depends on how you order them). So we no longer need to save the initial value

Reviewer: ```suggestion
be determined by checking if it's equal to one of the reserved slots.
Therefore we no longer need to save the initial value
```

> of `stack_ptr` in the cleanup phase.

Reviewer: ```suggestion
of `stack_ptr` in the cleanup phase. However, the frame pointer is now `alloc_ptr`-relative,
therefore we need to save the location of `prev_alloc_ptr` (since this was implicitly identical to `frame_ptr`).
```

> Thus we would have two different sets of reserved pointers in the execution and

Reviewer: ```suggestion
Thus, we have two different sets of reserved pointers between the execution and
```

cleanup phases:
 - `prev_frame_ptr`, `prev_stack_ptr`, `init_stack_ptr` in the execution phase,
   and
 - `prev_frame_ptr`, `prev_stack_ptr`, `prev_alloc_ptr` in the cleanup phase.
 
 Having which set of pointers are being saved change between phases is more
 confusing for the programmer. Thus, we put the `frame_ptr` in the "middle", and
 make the same `frame_ptr == stack_ptr` check to see whether the lightweight

>  stack is empty in both the execution and cleanup phases.

Reviewer: ```
This adds complexity for the programmer with no space or performance benefit to the system.
```
